docs/design/proposals/late-initialization/late-initialization.md

```diff
@@ -0,0 +1,113 @@
+
+This **late initialization** of AWS resource creates a difference between desired and latest state of k8s custom resource because currently the desired manifest does not get updated with these late initialized properties.
+
+This document will propose the solution for above problem and provide the details of implementing it.
+
+## Requirements
+
+A solution for this problem should :
+
+1. Update the desired state of the K8s custom resource with late initialized fields. 
+
+2. Provide configuration for AWS service teams to indicate which operation output returns the late initialized fields. i.e. Get/List/Create/Update output.
+
+    > AWS APIs have different behavior for returning the late initialized to the user. Some services return these fields as part of Create output, while some services will return the late initialized fields as part of  ReadOne output after the resource creation.
+
+3. Wait for asynchronous resource creation to populate late initialized fields.
+
+    > For some services/resources the late initialized fields do not appear immediately.  This happens when there is asynchronous processing involved in resource creation. The proposed solution should allow waiting and retry until the late initialized fields appear in the mentioned API output.
+
+4. Allow the AWS service teams to provide custom code using hooks for handing of late initialized fields.
+
+## Solution
+
+### TL;DR: 
+
+There will be a new `r.rd.AddLateInitializedFields()` method call added in the reconciliation loop, which will return `true` if the desired spec was updated with late initialized field(s) from the latest spec. If so, then reconciliation loop will patch the desired spec with new updates. 
+
+See the following sections for detailed explanation of this solution: 
+
+### Reconciler Updates
+
+* Create separate variables for `readOneLatest`, `createLatest` and `updateLatest` instead of a single `latest` variable which gets overwritten after readOne/Create/Update operation in the reconciliation loop.
+* Create a `desiredBase` which is copy of `desired` object after the status is updated(`patchResource()`) in etcd by the reconciler. We will use this copy as base to patch the desired resource spec in etcd.
+* Add `specUpdated, err := r.rd.AddLateInitializedFields(desired, readOneLatest, createLatest, updateLatest)`
+    > If the service teams have provided the configuration for populating late initialized field into `generator.yaml`, this method will update the desired spec accordingly and return `true, nil`. Otherwise it will return `false, nil`.
+* If `specUpdated` is `true`, reconciler will patch the resource in etcd using `desiredBase` as base object, otherwise no patching of K8s custom resource will take place and reconciliation will finish. 
+
```

Comment:
I'm not sure I'm no board with the above proposed solution. I was thinking that we would add a new `pkg/types.AWSResourceManager.SetOptionalFieldDefaults()` interface method and call that immediately after a successful call to `AWSResourceManager.Create()`.

```go
type AWSResourceManager interface {
...
    // SetDefaults determines if any optional fields in the supplied resource
    // have not been set by the user OR set by the resource manager in the response
    // from the Create output shape. If any such fields exist, SetDefaults calls
    // one or more AWS service API calls to retrieve the service-side default value of those
    // fields and returns a copy of the supplied resource with those default values set on the
    // resource's `Spec`.
    SetDefaults(AWSResource) (AWSResource, error)
}
```

In other words, instead of [this](https://github.com/aws-controllers-k8s/runtime/blob/089321f8f32690fab024fd0e633a0f70e9726262/pkg/runtime/reconciler.go#L232-L246):

```go
		rlog.Enter("rm.Create")
		latest, err = rm.Create(ctx, desired)
		rlog.Exit("rm.Create", err)
		if err != nil {
			if latest != nil {
				// this indicates, that even though Create failed
				// there is some changes available in the latest.RuntimeObject()
				// (example: ko.Status.Conditions) which have been
				// updated in the resource
				// Thus, patchResource() call should be made here
				_ = r.patchResource(ctx, desired, latest)
			}
			return err
		}
		rlog.Info("created new resource")
```

we would do something like this:

```go
		rlog.Enter("rm.Create")
		latest, err = rm.Create(ctx, desired)
		rlog.Exit("rm.Create", err)
		if err != nil {
			if latest != nil {
				// this indicates, that even though Create failed
				// there is some changes available in the latest.RuntimeObject()
				// (example: ko.Status.Conditions) which have been
				// updated in the resource
				// Thus, patchResource() call should be made here
				_ = r.patchResource(ctx, desired, latest)
			}
			return err
		}
		rlog.Enter("rm.SetDefaults")
		latestWithDefaults, err = rm.SetDefaults(ctx, latest)
		rlog.Exit("rm.SetDefaults", err)
		if err != nil {
			if latestWithDefaults != nil {
				// this indicates, that even though SetDefaults failed
				// there is some changes available in the latestWithDefaults.RuntimeObject()
				// (example: ko.Status.Conditions) which have been
				// updated in the resource
				// Thus, patchResource() call should be made here
				_ = r.patchResource(ctx, desired, latestWithDefaults)
			}
			return err
		}
		if rlog.DebugEnabled() {
			delta := r.rd.Delta(latest, latestWithDefaults)
			if delta.DifferentAt("Spec") {
				rlog.Debug("set default values on resource", "diff", delta.Differences)
			}
		}
		rlog.Info("created new resource")
```

Reply:
I see one concerns with the solution in comment:
**"SettingDefault" coupled with resource creation**
* This coupling can be problematic because I believe some APIs do not return late initialized fields as part of create response. Ex: `elasticache`, which does the asynchronous creation and the late initialized fields eventually showup as result of Get call after asynchronous creation has finished. 
* Another problem in the same vein is that `Create` resource call is only made once, if the default values are set as part of an `Update operation`, the above solution will not capture them. I can think of an example where customer just enables logging on their `apigateway stage` using `Update` operation, which sets the default value of `log level` to `Info`. If the logging is not enabled this value is not set.

Reply:
ACK Core decided to implement setting optional defaults field for the `Create` operation first, along with support for custom code.

Once this is implemented, ACK Core team will tackle `Update` and `ReadOne` operation separately as needed by the AWS service teams.